File: packages/dart/lib/src/vff.dart
import 'dart:typed_data';

import 'package:archive/archive.dart';

import 'errors.dart';

/// Container-level helpers for the modern (2021+) VFF/ZIP .skp format:
/// header validation, version-string extraction, and locating the embedded
/// ZIP payload. Ported from the steps in Python's _core.full_parse() that
/// run before TLV parsing begins.
class Vff {
  static final Uint8List _headerMagic =
      Uint8List.fromList([0xFF, 0xFE, 0xFF, 0x0E]);
  static final Uint8List _secondMarker = Uint8List.fromList([0xFF, 0xFE, 0xFF]);
  static final Uint8List _zipMagic =
      Uint8List.fromList([0x50, 0x4B, 0x03, 0x04]);

  static bool hasValidHeader(Uint8List data) {
    if (data.length < 4) return false;
    for (int i = 0; i < 4; i++) {
      if (data[i] != _headerMagic[i]) return false;
    }
    return true;
  }

  static int _findBytes(Uint8List data, Uint8List pattern, int start, int end) {
    if (pattern.isEmpty || start < 0) return -1;
    final limit = (end < data.length ? end : data.length) - pattern.length;
    for (int i = start < 0 ? 0 : start; i <= limit; i++) {
      bool match = true;
      for (int j = 0; j < pattern.length; j++) {
        if (data[i + j] != pattern[j]) {
          match = false;
          break;
        }
      }
      if (match) return i;
    }
    return -1;
  }

  /// Extract the "{17.0.18899}"-style version string from the
  /// UTF-16LE-encoded header region, or "unknown" if not found.
  static String extractVersion(Uint8List header) {
    final secondMarker = _findBytes(header, _secondMarker, 4, header.length);
    if (secondMarker <= 0) return 'unknown';

    final verStart = secondMarker + 4;
    if (verStart >= header.length) return 'unknown';

    String verText;
    try {
      final codeUnits = <int>[];
      for (int i = verStart; i + 1 < header.length; i += 2) {
        codeUnits.add(header[i] | (header[i + 1] << 8));
      }
      verText = String.fromCharCodes(codeUnits);
    } catch (_) {
      return 'unknown';
    }

    final braceStart = verText.indexOf('{');
    final braceEnd = verText.indexOf('}');
    if (braceStart >= 0 && braceEnd > braceStart) {
      return verText.substring(braceStart, braceEnd + 1);
    }
    return 'unknown';
  }

  /// Locate the byte offset of the embedded ZIP container ("PK\x03\x04"),
  /// searching only the first [searchLimit] bytes (matching Python's
  /// header-then-4096-byte-chunk search), or -1 if not found.
  static int findZipOffset(Uint8List data, [int searchLimit = 4096]) {
    final limit = searchLimit < data.length ? searchLimit : data.length;
    return _findBytes(data, _zipMagic, 0, limit);
  }

  static Archive openZip(Uint8List data, int zipOffset) {
    final zipBytes = Uint8List.sublistView(data, zipOffset);
    return ZipDecoder().decodeBytes(zipBytes);
  }

  // A ZIP entry's declared uncompressed size (ArchiveFile.size) is
  // untrusted central-directory metadata - it can be set independently of
  // what the compressed stream actually decompresses to, and even when
  // genuine, DEFLATE can expand highly compressible data by three orders
  // of magnitude. ArchiveFile.content decompresses (and so allocates) up
  // to that declared size with no ceiling of its own - decodeBytes() only
  // reads the central directory, so this check runs before any real
  // decompression happens. Real production model.dat entries are observed
  // at ~10x compression, so both limits below leave generous headroom for
  // legitimate files.
  static const int _maxUncompressedEntryBytes = 16 * 1024 * 1024 * 1024; // 16 GB
  static const int _maxCompressionRatio = 1000;
  static const int _ratioCheckThresholdBytes = 1024 * 1024; // 1 MB

  /// Reject a ZIP entry whose declared uncompressed size is implausible,
  /// before any code reads (and so decompresses/allocates for) its
  /// content.
  static void validateEntrySize(ArchiveFile entry) {
    final declared = entry.size;
    if (declared <= 0) return;

    if (declared > _maxUncompressedEntryBytes) {
      throw SkpParseException(
        "ZIP entry '${entry.name}' declares $declared bytes uncompressed, "
        'exceeding the $_maxUncompressedEntryBytes-byte safety ceiling',
        stage: 'zip_extract',
      );
    }

    if (declared >= _ratioCheckThresholdBytes) {
      final raw = entry.rawContent;
      final compressed = raw is ZipFile ? raw.compressedSize : 0;
      if (compressed <= 0 || declared / compressed > _maxCompressionRatio) {
        throw SkpParseException(
          "ZIP entry '${entry.name}' declares an implausible compression "
          "ratio ($declared bytes from $compressed bytes compressed) - "
          'likely a decompression bomb',
          stage: 'zip_extract',
        );
      }
    }
  }
}

File: packages/typescript/src/vff.ts
import { unzipSync, UnzipFileInfo } from 'fflate';

export interface SkpContents {
  version: string;
  modelData: Uint8Array;
  materialFiles: Record<string, Uint8Array>;
}

const VFF_MAGIC = [0xFF, 0xFE, 0xFF, 0x0E];
const ZIP_LOCAL_HEADER = [0x50, 0x4B, 0x03, 0x04]; // PK\x03\x04

function findSequence(data: Uint8Array, sequence: number[], startOffset: number = 0): number {
  const seqLen = sequence.length;
  if (seqLen === 0) return -1;
  const limit = data.length - seqLen;
  for (let i = startOffset; i <= limit; i++) {
    let match = true;
    for (let j = 0; j < seqLen; j++) {
      if (data[i + j] !== sequence[j]) {
        match = false;
        break;
      }
    }
    if (match) return i;
  }
  return -1;
}

export function validateHeader(data: Uint8Array): boolean {
  if (data.length < 4) return false;
  return (
    data[0] === VFF_MAGIC[0] &&
    data[1] === VFF_MAGIC[1] &&
    data[2] === VFF_MAGIC[2] &&
    data[3] === VFF_MAGIC[3]
  );
}

export function readVersion(data: Uint8Array): string {
  if (data.length < 16) return 'unknown';

  // Find second FF FE FF marker after the initial one at offset 0
  const secondMarker = findSequence(data, [0xFF, 0xFE, 0xFF], 4);
  if (secondMarker > 0) {
    const verStart = secondMarker + 4;
    const verBytes = data.subarray(verStart, Math.min(verStart + 200, data.length));
    try {
      const decoder = new TextDecoder('utf-16le');
      const verText = decoder.decode(verBytes);
      const braceStart = verText.indexOf('{');
      if (braceStart >= 0) {
        const braceEnd = verText.indexOf('}', braceStart);
        if (braceEnd > braceStart) {
          return verText.slice(braceStart, braceEnd + 1);
        }
      }
    } catch (e) {
      // Ignore decoder errors
    }
  }

  return 'unknown';
}

function findZipOffset(data: Uint8Array): number {
  const offset = findSequence(data, ZIP_LOCAL_HEADER);
  if (offset < 0) {
    throw new Error('No embedded ZIP archive found in the file');
  }
  return offset;
}

// A ZIP entry's declared uncompressed size (UnzipFileInfo.originalSize) is
// untrusted central-directory metadata - it can be set independently of
// what the compressed stream actually decompresses to, and even when
// genuine, DEFLATE can expand highly compressible data by three orders of
// magnitude. fflate's unzipSync decompresses (and so allocates) up to that
// declared size with no ceiling of its own. Real production model.dat
// entries are observed at ~10x compression, so both limits below leave
// generous headroom for legitimate files while rejecting the kind of
// declared-size lie or extreme ratio a genuine file would never need.
const MAX_UNCOMPRESSED_ENTRY_BYTES = 16 * 1024 * 1024 * 1024; // 16 GB
const MAX_COMPRESSION_RATIO = 1000;
const RATIO_CHECK_THRESHOLD_BYTES = 1024 * 1024; // 1 MB

/** Reject a ZIP entry whose declared uncompressed size is implausible,
 * before unzipSync decompresses (and so allocates for) it. */
function validateEntrySize(file: UnzipFileInfo): void {
  const declared = file.originalSize;
  if (declared <= 0) return;

  if (declared > MAX_UNCOMPRESSED_ENTRY_BYTES) {
    throw new Error(
      `ZIP entry '${file.name}' declares ${declared} bytes uncompressed, exceeding the ${MAX_UNCOMPRESSED_ENTRY_BYTES}-byte safety ceiling`
    );
  }

  if (declared >= RATIO_CHECK_THRESHOLD_BYTES) {
    const compressed = file.size;
    if (compressed <= 0 || declared / compressed > MAX_COMPRESSION_RATIO) {
      throw new Error(
        `ZIP entry '${file.name}' declares an implausible compression ratio (${declared} bytes from ${compressed} bytes compressed) - likely a decompression bomb`
      );
    }
  }
}

export function extractSkpContents(data: Uint8Array): SkpContents {
  // Allow both VFF-wrapped and bare ZIP (some exporters omit the header)
  if (!validateHeader(data)) {
    const zipInHeader = findSequence(data.subarray(0, Math.min(64, data.length)), ZIP_LOCAL_HEADER) >= 0;
    if (!zipInHeader) {
      throw new Error('Not a valid SketchUp (.skp) file');
    }
  }

  const version = readVersion(data);
  const zipOffset = findZipOffset(data);
  const zipBytes = data.subarray(zipOffset);

  // Only decompress entries we actually consume - a ZIP full of unrelated
  // large assets (the file's own thumbnails, etc.) would otherwise get
  // fully inflated into memory alongside model.dat for nothing.
  const wanted = (file: UnzipFileInfo): boolean => {
    const lower = file.name.toLowerCase();
    const isWanted = (
      lower === 'model.dat' ||
      lower.endsWith('/model.dat') ||
      lower.endsWith('.xml') ||
      lower.endsWith('.png') ||
      lower.endsWith('.jpg') ||
      lower.endsWith('.jpeg') ||
      lower.includes('material')
    );
    if (isWanted) validateEntrySize(file);
    return isWanted;
  };

  let unzipped: Record<string, Uint8Array>;
  try {
    unzipped = unzipSync(zipBytes, { filter: wanted });
  } catch (e) {
    throw new Error('Failed to decompress ZIP archive: ' + (e as Error).message);
  }

  let modelData: Uint8Array | null = null;
  const materialFiles: Record<string, Uint8Array> = {};

  for (const entry of Object.keys(unzipped)) {
    const lower = entry.toLowerCase();
    if (lower === 'model.dat' || lower.endsWith('/model.dat')) {
      modelData = unzipped[entry];
    } else if (
      lower.endsWith('.xml') ||
      lower.endsWith('.png') ||
      lower.endsWith('.jpg') ||
      lower.endsWith('.jpeg') ||
      lower.includes('material')
    ) {
      materialFiles[entry] = unzipped[entry];
    }
  }

  if (!modelData) {
    throw new Error('ZIP archive found but does not contain a model.dat entry');
  }

  return {
    version,
    modelData,
    materialFiles,
  };
}
